py/bitbox02/bitbox02/communication/bitbox_api_protocol.py

```diff
@@ -270,21 +301,133 @@ def _create_noise_channel(self, noise_config: BitBoxNoiseConfig) -> NoiseConnect
+        """
+        V1-6 of the BB noise protocol doesn't encapsulate handshake requests, and don't
+        send back a status code in the response.
+        """
+        noise_result = self._raw_query(req)
+        return RESPONSE_SUCCESS, noise_result
+
+class BitBoxProtocolV2(BitBoxProtocolV1):
+    """ BitBox Protocol from firmware V2.0.0 onwards. """
+
+    def unlock_query(self) -> bytes:
+        unlock_result, unlock_data = self.query(OP_UNLOCK, b"")
+        if len(unlock_data) != 0:
+            raise ValueError(f"OP_UNLOCK (V3) replied with wrong length.")
+        return unlock_result
+
+
+class BitBoxProtocolV3(BitBoxProtocolV2):
+    """ BitBox Protocol from firmware V3.0.0 onwards. """
+
+    def unlock_query(self) -> bytes:
+        unlock_result, unlock_data = self.query(OP_UNLOCK, b"")
+        if len(unlock_data) != 0:
+            raise ValueError(f"OP_UNLOCK (V3) replied with wrong length.")
+        return unlock_result
+
+
+class BitBoxProtocolV4(BitBoxProtocolV3):
+    """ BitBox Protocol from firmware V4.0.0 onwards. """
+
+    def _encode_noise_request(self, encrypted_msg: bytes) -> bytes:
+        return OP_NOISE_MSG + encrypted_msg
+
+
+class BitBoxProtocolV7(BitBoxProtocolV4):
+    """ Noise Protocol from firmware V7.0.0 onwards. """
+
+    def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
+        return self.query(OP_HER_COMEZ_TEH_HANDSHAEK, req)
+
+    def _decode_noise_response(self, encrypted_msg: bytes) -> Tuple[bytes, bytes]:
+        print("Decoding: " + encrypted_msg.hex())
+        return encrypted_msg[0:1], encrypted_msg[1:]
+
+
+class BitBoxCommonAPI:
+    """Class to communicate with a BitBox device"""
+
+    # pylint: disable=too-many-public-methods,too-many-arguments
+    def __init__(
+        self, transport: TransportLayer, device_info: DeviceInfo, noise_config: BitBoxNoiseConfig
+    ):
+        self.debug = False
+        serial_number = device_info["serial_number"]
+
+        self.version = parse_device_version(serial_number)
+        if self.version is None:
+            transport.close()
+            raise ValueError(f"Could not parse version from {serial_number}")
+        # Delete the prelease part, as it messes with the comparison (e.g. 3.0.0-pre < 3.0.0 is
+        # True, but the 3.0.0-pre has already the same API breaking changes like 3.0.0...).
+        self.version = semver.VersionInfo(
+            self.version.major, self.version.minor, self.version.patch, build=self.version.build
+        )
+
+        self._bitbox_protocol: BitBoxProtocol
+        if self.version >= semver.VersionInfo(7, 0, 0):
+            self._bitbox_protocol = BitBoxProtocolV7(transport)
+        elif self.version >= semver.VersionInfo(4, 0, 0):
+            self._bitbox_protocol = BitBoxProtocolV4(transport)
+        elif self.version >= semver.VersionInfo(3, 0, 0):
+            self._bitbox_protocol = BitBoxProtocolV3(transport)
+        else:
```

Comment:
BitBoxProtocolV2 missing (and if it is there, it runs into the `Unlock process aborted` exception)

Reply:
Fixed in #362